go/libraries/doltcore/doltdb/root_val.go

```diff
@@ -188,11 +188,34 @@ func (root *RootValue) GenerateTagsForNewColColl(ctx context.Context, tableName
 
 // GenerateTagsForNewColumns deterministically generates a slice of new tags that are unique within the history of this root. The names and NomsKinds of
 // the new columns are used to see the tag generator.
-func (root *RootValue) GenerateTagsForNewColumns(ctx context.Context, tableName string, newColNames []string, newColKinds []types.NomsKind) ([]uint64, error) {
+func (root *RootValue) GenerateTagsForNewColumns(ctx context.Context, tableName string, newColNames []string, newColKinds []types.NomsKind, headRoot *RootValue) ([]uint64, error) {
 	if len(newColNames) != len(newColKinds) {
 		return nil, fmt.Errorf("error generating tags, newColNames and newColKinds must be of equal length")
 	}
 
+	if headRoot != nil {
+		tbl, found, err := headRoot.GetTable(ctx, tableName)
+		if err != nil {
+			return nil, err
+		}
+
+		if found {
+			sch, err := tbl.GetSchema(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			colOverlap := schema.GetSharedCols(sch, newColNames, newColKinds)
+			if len(colOverlap) == len(newColNames) {
+				var reusedTags []uint64
+				for _, k := range newColNames {
+					reusedTags = append(reusedTags, colOverlap[k])
+				}
+				return reusedTags, nil
+			}
```

Comment:
@zachmu lmk if this is similar to what you wanted. this has the added benefit of extending to partial re-use later